cmd/tether/ops_linux.go

```diff
@@ -177,33 +178,85 @@ func (t *operations) SetupFirewall(config *tether.ExecutorConfig) error {
 			ifaceName := iface.Attrs().Name
 			log.Debugf("slot %d -> %s", endpoint.ID, ifaceName)
 
-			established.Interface = ifaceName
-			_ = established.Commit(context.TODO())
-
-			// handle the ports
-			for _, p := range endpoint.Ports {
-				// parse the port maps
-				r, err := portToRule(p)
-				if err != nil {
-					log.Errorf("can't apply port rule (%s): %s", p, err.Error())
-					continue
+			switch endpoint.Network.TrustLevel {
+			case common.Open:
+				// Accept all incoming and outgoing traffic
+				generalPolicy(netfilter.Accept)
+
+			case common.Closed:
+				// Reject all incoming and outgoing traffic
+				generalPolicy(netfilter.Drop)
+
+			case common.Outbound:
+				// Reject all incoming traffic, but allow outgoing
+				setupOutboundFirewall(ifaceName)
+
+			case common.Peers:
+				// Outbound + all ports open to source addresses in --container-network-ip-range
+				setupOutboundFirewall(ifaceName)
+				sourceAddresses := make([]string, len(endpoint.Network.Pools))
+				for i, v := range endpoint.Network.Pools {
+					sourceAddresses[i] = v.String()
+				}
+				(&netfilter.Rule{
+					Chain:           netfilter.Input,
+					Target:          netfilter.Accept,
+					SourceAddresses: sourceAddresses,
+					Interface:       ifaceName,
+				}).Commit(context.TODO())
+
+			case common.Published:
+				established.Interface = ifaceName
+				_ = established.Commit(context.TODO())
+				// handle the ports
+				for _, p := range endpoint.Ports {
+					// parse the port maps
+					r, err := portToRule(p)
+					if err != nil {
+						log.Errorf("can't apply port rule (%s): %s", p, err.Error())
+						continue
+					}
+
+					log.Infof("Applying rule for port %s", p)
+					r.Interface = ifaceName
+					_ = r.Commit(context.TODO())
 				}
+				reject.Interface = ifaceName
+				_ = reject.Commit(context.TODO())
 
-				log.Infof("Applying rule for port %s", p)
-				r.Interface = ifaceName
-				_ = r.Commit(context.TODO())
+			default:
+				log.Warning("Received invalid firewall configuration %v: defaulting to open.",
+					endpoint.Network.TrustLevel)
+				generalPolicy(netfilter.Accept)
 			}
-
-			reject.Interface = ifaceName
-			_ = reject.Commit(context.TODO())
-
 			break
 		}
 	}
 
 	return nil
 }
 
+func setupOutboundFirewall(ifaceName string) {
+	netfilter.Policy(context.TODO(), netfilter.Input, netfilter.Drop)
```

Comment:
I'm unsure whether these rules will be appended or inserted and couldn't quickly find whether this is using `iptables-restore` to atomically perform the changed.

There are various operations that cause the tether to reapply it's configuration - the philosophy is "idempotence" so that isn't a problem in general, but I cannot confirm that we don't slowly build up more and more rules. Please run 'exec' against a container and confirm that:
a. the tether.debug shows the rules being checked/applied a second time if needed
b. the iptables state is sane afterwards and we're not appending a new set of duplicate rules each time.